### PR TITLE
Workaround to reduce the stack usage

### LINKED ISSUE #1960: Excessive stack usage in setup() due to font setup
**Operating environment/Installation (Hass.io/Docker/pip/etc.):**

pip under Debian Linux

**ESP (ESP32/ESP8266, Board/Sonoff):**

ESP32 Devkit

**ESPHome version (latest production, beta, dev branch)**

dev branch

**Affected component:**

code generation

**Description of problem:**
Larger setups tend to crash, see e.g. #855 for the symptoms. The culprit seem to be an excessive stack usage of the generated setup() function.

I am including the output for `-fno-inline -Wstack-usage=256`. The `setup()` is indeed a clear winner with a 3120 bytes of stack (3248 if inlining is enabled so inlined methods are not the problem here). That did not make much sense to me as pretty much everything put on stack there are the negligible colors and `wifi::WiFiAP` (with sizeof 112). 

After some more digging my opinion is that the stack usage is connected to the font creation generated as

```
font1 = new display::Font({display::Glyph(" ", uint8_t_2, 0, 0, 15, 5, 0), display::Glyph("!", uint8_t_2, 0, 0, 3, 6, 12), ...
```
I am not an expert on C++ compiler details but it is the _only_ place that smells of stack usage. I changed the font to include just one glyph which immediately brought the stack usage to 1424 bytes. So in my opinion this should be better done somehow iteratively.

This is probably _not_ the reason for #855 as fonts do not seem to be mentioned there. There might be however other pieces of generated `{ foo, bar, ... }` code prone to a similar problem.

Also note the `void esphome::ota::OTAComponent::handle_()` with 1248 bytes that indeed puts a char[1024] on stack; that should be probably fixed but unlikely to kick in here. Quite a few 3rd party methods are flagged with a potentially unbounded usage mostly doing `char foo[strlen(something)...]` but neither seems to likely ask for too much in the common scenarios.

**Problem-relevant YAML-configuration entries:**
```yaml
font:
  - file: "fonts/DejaVuSans.ttf"
    id: font1
    size: 16
    # glyphs: [ '0' ]
```

**Logs (if applicable):**

```
src/esphome/components/display/display_buffer.cpp: In member function 'void esphome::display::DisplayBuffer::vprintf_(int, int, esphome::display::Font*, esphome::Color, esphome::display::TextAlign, const char*, va_list)':
src/esphome/components/display/display_buffer.cpp:201:6: warning: stack usage is 336 bytes [-Wstack-usage=]
 void DisplayBuffer::vprintf_(int x, int y, Font *font, Color color, TextAlign align, const char *format, va_list arg) {
      ^
src/esphome/components/ota/ota_component.cpp: In member function 'void esphome::ota::OTAComponent::handle_()':
src/esphome/components/ota/ota_component.cpp:51:6: warning: stack usage is 1248 bytes [-Wstack-usage=]
 void OTAComponent::handle_() {
      ^
src/esphome/components/wifi/wifi_component_esp32.cpp: In member function 'bool esphome::wifi::WiFiComponent::wifi_sta_connect_(esphome::wifi::WiFiAP)':
src/esphome/components/wifi/wifi_component_esp32.cpp:152:6: warning: stack usage is 336 bytes [-Wstack-usage=]
 bool WiFiComponent::wifi_sta_connect_(WiFiAP ap) {
      ^
src/esphome/components/wifi/wifi_component.cpp: In member function 'void esphome::wifi::WiFiComponent::check_scanning_finished()':
src/esphome/components/wifi/wifi_component.cpp:329:6: warning: stack usage is 288 bytes [-Wstack-usage=]
 void WiFiComponent::check_scanning_finished() {
      ^
src/main.cpp: In function 'void setup()':
src/main.cpp:92:6: warning: stack usage is 3120 bytes [-Wstack-usage=]
 void setup() {
      ^
/home/stano/.platformio/packages/framework-arduinoespressif32@3.10004.210126/libraries/WiFi/src/WiFiAP.cpp: In member function 'bool WiFiAPClass::softAP(const char*, const char*, int, int, int)':
/home/stano/.platformio/packages/framework-arduinoespressif32@3.10004.210126/libraries/WiFi/src/WiFiAP.cpp:92:6: warning: stack usage is 288 bytes [-Wstack-usage=]
 bool WiFiAPClass::softAP(const char* ssid, const char* passphrase, int channel, int ssid_hidden, int max_connection)
      ^
/home/stano/.platformio/packages/framework-arduinoespressif32@3.10004.210126/libraries/WiFi/src/WiFiSTA.cpp: In member function 'wl_status_t WiFiSTAClass::begin(const char*, const char*, int32_t, const uint8_t*, bool)':
/home/stano/.platformio/packages/framework-arduinoespressif32@3.10004.210126/libraries/WiFi/src/WiFiSTA.cpp:115:13: warning: stack usage is 288 bytes [-Wstack-usage=]
 wl_status_t WiFiSTAClass::begin(const char* ssid, const char *passphrase, int32_t channel, const uint8_t* bssid, bool connect)
             ^
/home/stano/.platformio/packages/framework-arduinoespressif32@3.10004.210126/libraries/ESPmDNS/src/ESPmDNS.cpp: In member function 'void MDNSResponder::enableWorkstation(wifi_interface_t)':
/home/stano/.platformio/packages/framework-arduinoespressif32@3.10004.210126/libraries/ESPmDNS/src/ESPmDNS.cpp:114:6: warning: stack usage might be unbounded [-Wstack-usage=]
 void MDNSResponder::enableWorkstation(wifi_interface_t interface){
      ^
/home/stano/.platformio/packages/framework-arduinoespressif32@3.10004.210126/libraries/ESPmDNS/src/ESPmDNS.cpp: In member function 'void MDNSResponder::addService(char*, char*, uint16_t)':
/home/stano/.platformio/packages/framework-arduinoespressif32@3.10004.210126/libraries/ESPmDNS/src/ESPmDNS.cpp:133:6: warning: stack usage might be unbounded [-Wstack-usage=]
 void MDNSResponder::addService(char *name, char *proto, uint16_t port){
      ^
/home/stano/.platformio/packages/framework-arduinoespressif32@3.10004.210126/libraries/ESPmDNS/src/ESPmDNS.cpp: In member function 'bool MDNSResponder::addServiceTxt(char*, char*, char*, char*)':
/home/stano/.platformio/packages/framework-arduinoespressif32@3.10004.210126/libraries/ESPmDNS/src/ESPmDNS.cpp:152:6: warning: stack usage might be unbounded [-Wstack-usage=]
 bool MDNSResponder::addServiceTxt(char *name, char *proto, char *key, char *value){
      ^
/home/stano/.platformio/packages/framework-arduinoespressif32@3.10004.210126/libraries/ESPmDNS/src/ESPmDNS.cpp: In member function 'int MDNSResponder::queryService(char*, char*)':
/home/stano/.platformio/packages/framework-arduinoespressif32@3.10004.210126/libraries/ESPmDNS/src/ESPmDNS.cpp:190:5: warning: stack usage might be unbounded [-Wstack-usage=]
 int MDNSResponder::queryService(char *service, char *proto) {
     ^
.piolibdeps/gui/ESPAsyncWebServer-esphome/src/WebAuthentication.cpp: In function 'bool checkDigestAuthentication(const char*, const char*, const char*, const char*, const char*, bool, const char*, const char*, const char*)':
.piolibdeps/gui/ESPAsyncWebServer-esphome/src/WebAuthentication.cpp:144:6: warning: stack usage is 384 bytes [-Wstack-usage=]
 bool checkDigestAuthentication(const char * header, const char * method, const char * username, const char * password, const char * realm, bool passwordIsHash, const char * nonce, const char * opaque, const char * uri){
      ^
.piolibdeps/gui/ESPAsyncWebServer-esphome/src/WebHandlers.cpp: In member function 'virtual void AsyncStaticWebHandler::handleRequest(AsyncWebServerRequest*)':
.piolibdeps/gui/ESPAsyncWebServer-esphome/src/WebHandlers.cpp:187:6: warning: stack usage is 272 bytes [-Wstack-usage=]
 void AsyncStaticWebHandler::handleRequest(AsyncWebServerRequest *request)
      ^
.piolibdeps/gui/ESPAsyncWebServer-esphome/src/WebRequest.cpp: In member function 'bool AsyncWebServerRequest::_parseReqHeader()':
.piolibdeps/gui/ESPAsyncWebServer-esphome/src/WebRequest.cpp:312:6: warning: stack usage is 320 bytes [-Wstack-usage=]
 bool AsyncWebServerRequest::_parseReqHeader(){
      ^
.piolibdeps/gui/ESPAsyncWebServer-esphome/src/WebResponses.cpp: In member function 'virtual String AsyncWebServerResponse::_assembleHead(uint8_t)':
.piolibdeps/gui/ESPAsyncWebServer-esphome/src/WebResponses.cpp:126:8: warning: stack usage might be 400 bytes [-Wstack-usage=]
 String AsyncWebServerResponse::_assembleHead(uint8_t version){
        ^
.piolibdeps/gui/ESPAsyncWebServer-esphome/src/WebResponses.cpp: In constructor 'AsyncFileResponse::AsyncFileResponse(AsyncFileResponse::FS&, const String&, const String&, bool, AwsTemplateProcessor)':
.piolibdeps/gui/ESPAsyncWebServer-esphome/src/WebResponses.cpp:507:1: warning: stack usage might be unbounded [-Wstack-usage=]
 AsyncFileResponse::AsyncFileResponse(FS &fs, const String& path, const String& contentType, bool download, AwsTemplateProcessor callback): AsyncAbstractResponse(callback){
 ^
.piolibdeps/gui/ESPAsyncWebServer-esphome/src/WebResponses.cpp: In constructor 'AsyncFileResponse::AsyncFileResponse(AsyncFileResponse::File, const String&, const String&, bool, AwsTemplateProcessor)':
.piolibdeps/gui/ESPAsyncWebServer-esphome/src/WebResponses.cpp:541:1: warning: stack usage might be unbounded [-Wstack-usage=]
 AsyncFileResponse::AsyncFileResponse(File content, const String& path, const String& contentType, bool download, AwsTemplateProcessor callback): AsyncAbstractResponse(callback){
 ^
/home/stano/.platformio/packages/framework-arduinoespressif32@3.10004.210126/cores/esp32/Esp.cpp: In function 'uint32_t sketchSize(sketchSize_t)':
/home/stano/.platformio/packages/framework-arduinoespressif32@3.10004.210126/cores/esp32/Esp.cpp:144:17: warning: stack usage is 304 bytes [-Wstack-usage=]
 static uint32_t sketchSize(sketchSize_t response) {

```

**Additional information and things you've tried:**

### PR DESCRIPTION
**Pull request by numo68**
# What does this implement/fix? 

A proof of concept of the way to lower the stack usage of `setup()`

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Configuration change (this will require users to update their yaml configuration files to keep working)

**Related issue or feature (if applicable):** fixes esphome/issues#1960
  
# Test Environment

- [x] ESP32
- [ ] ESP8266
- [ ] Windows
- [ ] Mac OS
- [ ] Linux

# Explain your changes

This is a demonstration of an admittedly ugly workaround for excessive stack usage in `setup()`. Doing googling and some trial and error I found that the following way can be used to force a piece of code to localize its stack usage by basically forcing it out to an own function

```cpp
[&]() __attribute__((noinline)) {
  int table[1000];
  f(table, 1000);
}();
```

My quite complicated project started with the stack usage of `setup()` of 14544 bytes. Adjusting `CONFIG_ARDUINO_LOOP_STACK_SIZE` for the whole thing not to crash caused it to run out of memory, manifesting as occasional crashes and reproducible inability to do OTA updates.

The workaround is activated for the three constructs
- the font creation (brought the `setup()` to 12720 bytes with largest lambda 1888)
- a crude attempt to catch initializer lists without catching anything that hurts (9408 bytes)
- the lambdas containing `[=](` (8944 bytes)

That means that the total, i.e. `setup()` plus largest font creation lambda is now 10832 bytes, a saving of 3.7 kB. Still needs adjusting the `sdkconfig.h` but saves a bit.

I am of course not calling for merging it, it is just a proof of concept and maybe a help for users affected.

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/core/__init__.py [hunk 1]
```diff
@@ -712,13 +717,21 @@ def register_variable(self, id, obj):
     def has_id(self, id):
         return id in self.variables
 
+    def reduce_stack_usage(self, stmt):
+        if self.stack_reduction_not_re.search(stmt) is not None:
+            return stmt
+        if self.stack_reduction_re.search(stmt) is not None:
+            return "[&]() __attribute__((noinline)) {" + stmt + "}();"
```
Comment (OttoWinter) [inline esphome/core/__init__.py:724]:
```suggestion
            return "[&]() __attribute__((noinline)) {" + stmt + "}();"
```

Hmm, if this replaced a lambda like `bar([=]() { my_local_var.foo(); })` where `my_local_var` is a local variable in the `setup` function I think this might not work.

Because then it will be replaced by

```cpp
[&]() __attribute__((noinline)) { 
  bar([=]() { my_local_var.foo(); });
}();
```

but the outer lambda uses the reference, so the inner one is (I think) illegal.

Anyway, the `[=]` lambda capture was only necessary in very early versions anyway (where all variables were locally defined). I think we can just replace the scope on the normal lambda expression.

File: esphome/core/__init__.py [hunk 2]
```diff
@@ -489,6 +489,11 @@ def __init__(self):
         self.component_ids = set()
         # Whether ESPHome was started in verbose mode
         self.verbose = False
+        # Regexps to search for stack usage reduction workaround
+        self.stack_reduction_re = re.compile(
+            r"(\([a-zA-Z_\d ]*\{[a-zA-Z_\d, {}]*\)\;)|(\[\=\]\()"
```
Comment (OttoWinter) [inline esphome/core/__init__.py:494]: I know it's been some time (sorry about that!), but could you add a few comments here what this is supposed to match?

Like a short description how the regex generally works, plus some small examples.

Reason: I tried to understand it for a bit now, but can't wrap my head around it. It will be a nightmare to maintain

Comment (numo68) [reply esphome/core/__init__.py:494]: Lol... okay, I need to rethink this as it is indeed quite some time. Sadly I did not make any notices for me either. I'll come back when I have more time. Frankly I think I am seeing some weirdness there - I don't think the `[a-zA-Z_\d ]*` is ever matched.

The main point of this all was to catch the initializer lists of the form like

```
binary_sensor_autorepeatfilter_3 = new binary_sensor::AutorepeatFilter({{500, 300, 200}, {3000, 125, 125}});
```

that result in the unnecessary stack usage. I.e. the goal here is mainly to pick a `(` followed by `{`, followed by alphanumeric, space, comma or braces and ending with `);`. The second part catches any `[=](` sequence. The negative regex... well, no idea, it reverted some unwanted match but I have no idea what it was :-D

But as said, let me look at this later and come back to you.

It is just a proof of concept and it is very well possible that there are edge cases. It however compiled everything in the test folder.
